website/api/zkvm/input-output.md

```diff
@@ -0,0 +1,274 @@
+- [`env::write_slice`][write_slice_method]: Writes a slice to the `stdout` file
+  descriptor.
+
+Differently from the [`stdin`][stdin] file descriptor, in the [`stdout`][stdout]
+we need to provide a buffer to write the data to.
+
+The basic flow is that we need to instantiate a buffer in the host program, and
+then pass it to the guest program. The guest program will write the output to
+the [`stdout`][stdout] file descriptor, and the host program will read the
+output from the buffer.
+
+For example, in the host program, we could:
+
+`/src/main.rs`
+
+```rust
+use risc0_zkvm::{
+  ExecutorEnv,
+  Receipt,
+  default_prover,   // convenience method to create a default prover
+  serde::from_slice // necessary to deserialize the output
+};
+use app_methods::EXAMPLE_ELF;
+use app::ExampleStruct;
+
+fn main() {
+  // Step 1: Create a buffer to store the output
+  let mut output: Vec<u8> = Vec::new();
+  let env = ExecutorEnv::builder()
+              // Step 2: Instruct the guest to write to the desired buffer
+              .stdout(&mut output)
+              .build();
+  let prover = default_prover();
+  let receipt = prover.prove(env, EXAMPLE_ELF).unwrap();
+
+  // Step 3: Deserialize the output
+  let example_struct: ExampleStruct = from_slice(&output).unwrap();
+  let example_slice: [u8; 3] = from_slice(&output).unwrap();
+
+  // Do something with the output...
+}
+```
+
+And in the guest program, we could:
+
+`/methods/guest/src/main.rs`
+
+```rust
+#![no_main]
+use risc0_zkvm::guest::env;
+risc0_zkvm::guest::entry!(main);
+use app::ExampleStruct;
+
+fn main() {
+  // Write a single value to [stdout]
+  env::write(ExampleStruct {
+    field1: 42,
+    field2: "Hello, World!".to_string()
+  });
+
+  // Write a slice to [stdout]
+  env::write_slice(&[1u8, 2u8, 3u8]);
+
+  // Do more things...
+}
+```
+
+## Handling Public Outputs
+
+The [`journal`][journal] file descriptor is used to send public output. This
+will be part of the proof and will be included in the journal. The zkVM provides
+some convenience methods to write data to this file descriptor.
+
+- [`env::commit`][commit_method]: Writes a value to the `journal` file descriptor.
+- [`env::commit_slice`][commit_method]: Writes a slice to the `journal` file descriptor.
+
+Differently from the [`stdout`][stdout] file descriptor, in the
+[`journal`][journal] we just need to provide the data to be written, no buffer
+is necessary.
+
+For example, in the host program, we could instantiate the `ExecutorEnv` as usual, and then get the journal data from the receipt.
+
+`/src/main.rs`
+
+```rust
+use risc0_zkvm::{default_prover, ExecutorEnv};
+
+fn main() {
+  let env = ExecutorEnv::builder()
+              .build();
+  let prover = default_prover();
+  let receipt = prover.prove(env, EXAMPLE_ELF).unwrap();
+
+  // Get the journal data from the receipt
+  let journal = receipt.journal;
+
+  // Do something with the journal data...
```

Comment:
Like the examples you have above, it would be good to show decoding the journal data. In particular, the `decode()` function on the `Journal` type is a nice convenience to show devs.